get_indexer_non_unique for orderable indexes

horta opened the pull request:
Tackles the performance issue raised at #15364 by looping simultaneously over source and target indexes, avoiding dictionary and set lookups in for/while loops.

I guess it can be further improved in terms of static typing. For example, if both source and target indexes are simple C types, the function `_indexer_non_unique_orderable_loop` should be way faster if we provide that information.

jreback committed:
TST: use xdist for multiple cpu testing

closes #15369

horta committed:
get_indexer_non_unique for orderable indexes

horta committed:
get_indexer_non_unique for orderable indexes

horta committed:
get_indexer_non_unique for orderable indexes

C level list

no gil

capture index error

wrong exception handling

jreback commented on pandas/indexes/base.py:
```diff
@@ -2509,7 +2509,18 @@ def get_indexer_non_unique(self, target):
 
-        indexer, missing = self._engine.get_indexer_non_unique(tgt_values)
+        try:
+            if self.is_all_dates:
```
you just need to check ``is_monotonic_increasing`` to see if you can do this.

jreback commented:
using lists here is not really great (not that the original is much better).

jreback commented:
need an asv to see if this is actually better

jreback commented:
prob a better soln here.

factorize both the values and the indexers (so you have int64). then sort these.

shoyer commented on pandas/index.pyx:
```diff
@@ -45,6 +45,47 @@ cdef extern from "Python.h":
+cdef _indexer_non_unique_orderable_loop(ndarray values, ndarray targets,
+                                        int64_t[:] idx0,
+                                        int64_t[:] idx1,
+                                        list[:] result, list[:] missing):
```
`Int64Vector` from https://github.com/pandas-dev/pandas/blob/master/pandas/src/hashtable_class_helper.pxi.in would be a much faster way to handle accumulating these arrays.

neirbowj committed:
BUG: Avoid grafting missing examples directory (#15373)

jreback committed:
CLN: remove pandas/io/auth.py, from ga.py (now removed) (#15374)

jreback committed:
TST: consolidate remaining tests under pandas.tests

move all remaining tests so that ALL tests are now
under pandas/tests

Author: Jeff Reback <jeff@reback.net>

Closes #15371 from jreback/tests and squashes the following commits:

43039e4 [Jeff Reback] add in data
118127b [Jeff Reback] wip
bfa6a9c [Jeff Reback] fix data locations
79a79e6 [Jeff Reback] fix import
57437bf [Jeff Reback] fixes
b407586 [Jeff Reback] move io
e13bfe3 [Jeff Reback] move tools
0194e31 [Jeff Reback] move computation
0e6bcb4 [Jeff Reback] rename test_msgpack -> msgpack
c5e4ab8 [Jeff Reback] move sparse
42e60e2 [Jeff Reback] move api tests

jreback committed:
TST: fix locations for github based url tests

jreback committed:
DOC: fix path in whatsnew

jreback committed:
TST: control skipping of numexpr tests if its installed / used

andrewkittredge committed:
Typo (#15377)

jreback committed:
TST: make test_gbq single cpu

horta committed:
C level list

horta committed:
C level list

Horta commented:
The last commit speed up quite a lot:

```python
import numpy as np
import pandas as pd
from pandas import Index

from IPython import get_ipython
ipython = get_ipython()

i = Index(Index(np.arange(1000000)).tolist() + Index(np.arange(1000000)).tolist())

ipython.magic("timeit i.get_indexer_non_unique(i[0:1000000])")
```

Previous implementation: 1 loop, best of 3: 3.42 s per loop
Pull request implementation: 1 loop, best of 3: 681 ms per loop

horta committed:
no gil

horta committed:
no gil

shoyer commented:
@Horta Can you simply reuse `Int64Vector` instead of making a new `Int64List` type?

Horta commented:
@shoyer I've tried and couldn't make it work. Also, I don't thing contiguous array is the way to go for such a problem as some indices might not repeat, or might repeat a lot.

But I'm working on another implementation that relies only on fixed-size arrays, which means that I won't need any of those two types (Int64List or Int64Vector).

shoyer commented:
>  I've tried and couldn't make it work

Can you clarify? You couldn't get it to compile or you couldn't understand how to use it?

> Also, I don't thing contiguous array is the way to go for such a problem as some indices might not repeat, or might repeat a lot.

Couldn't we simply accumulative the indexes of matching elements, e.g., by repeatedly appending to a Python list of integers? That's effectively what `Int64Vector` is for, except without the overhead of boxing in Python objects. It's essentially a C++ vector written in Cython.

I would rather not add new low level data structure for this specific method unless we absolutely need it. My understanding is that vectors are usually faster than linked lists, anyways. It's certainly more memory efficient.

Horta commented:
> Can you clarify? You couldn't get it to compile or you couldn't understand how to use it?

I could not import it. Looks like it is private (only in pyx). I tried to export defining them on pxd and had some difficults. Then I tried to copy/paste Int64Vector and still had difficults. Basically it needs documentation...

> I would rather not add new low level data structure for this specific method unless we absolutely need it.

The implementation I'm working on makes no use of new types: I will be using basic C types. It is a matter of looping over the src and tgt values twice, calling Malloc twice and realloc once.

shoyer commented:
Take a look at the IntervalIndex PR: https://github.com/pandas-dev/pandas/pull/15309/files#diff-cdfa55be00a117a8466e28b72f60cfca

You should be able to just copy the lines from that change for pandas/hashtable.pxd and then use `from hashtable cimport Int64Vector`

jreback commented:
I may acutally push some of the low-level code into master later / tomrrow if I get a chance to make this easier anyhow.

jreback committed:
ENH: expose Int64VectorData in hashtable.pxd

jreback committed:
TST: xfail most test_gbq tests for now

jreback commented:
I just pushed these to master: https://github.com/pandas-dev/pandas/commit/010393c4cb650b78e3e51af417e7037737e8d3b6

horta committed:
capture index error

horta committed:
capture index error

horta committed:
wrong exception handling

horta committed:
wrong exception handling

parthea committed:
TST: Fix gbq integration tests. gbq._Dataset.dataset() would not return full results

This PR resolves an issue where `gbq._Dataset.datasets()` would not return
all datasets under a Google BigQuery project.

If `'nextPageToken'` is populated, then another `datasets().list()`
request should be sent with `'pageToken'` set to collect more results.
In the past few days, additional datasets were added under the Google
BigQuery project id used by pandas as part of the following github
project : https://github.com/pydata/pandas-gbq .

The addition of datasets caused many gbq unit tests to fail because in function
`clean_gbq_environment()`, we check to see if the dataset exists using
the incomplete results from `gbq._Dataset.datasets()` before we
attempt to delete it.

Author: Anthonios Partheniou <apartheniou@electricalengineer.ca>

Closes #15381 from parthea/fix-broken-gbq-unit-tests and squashes the following commits:

61bc1e7 [Anthonios Partheniou] TST: Fix gbq tests. gbq.dataset()/gbq.tables would not return full results.

mroeschke committed:
Bug: Raise ValueError with interpolate & fillna limit = 0 (#9217)

closes #9217

Author: Matt Roeschke <emailformattr@gmail.com>

Closes #14994 from mroeschke/fix_9217 and squashes the following commits:

c1790ee [Matt Roeschke] Unify ValueError message and correct cython limits
6f041e6 [Matt Roeschke] Bug: Raise ValueError with interpolate limit = 0

jreback committed:
CLN: create core/sorting.py

just a small reorg to put sorting / grouping utilities into a separate
area

Author: Jeff Reback <jeff@reback.net>

Closes #15402 from jreback/sorting and squashes the following commits:

fdcf9a1 [Jeff Reback] change a couple of sorting.py functions to be non-private (public to pandas internals)
90ff22d [Jeff Reback] split up some value_counts groupby tests a bit
18ea902 [Jeff Reback] CLN: create core/sorting.py
92dcb07 [Jeff Reback] CLN: remove numpy_groupby as not used

jreback committed:
TST: disable gbq tests again

jreback committed:
TST: fix incorrect url in compressed url network tests in parser

jreback committed:
TST: incorrect skip in when --skip-network is run

closes #15407

jreback committed:
TST: fix test_nework.py fixture under py27

Commit:
BLD: Numexpr 2.4.6 required

closes #15213

Author: Francesc Alted <francesc@continuum.io>

Closes #15383 from FrancescAlted/numexpr-2.4.6 and squashes the following commits:

c417fe2 [Francesc Alted] Simplify and remove UserWarning testing on numexpr import
e1b34a9 [Francesc Alted] Force a reload of pd.computation for actually triggering the UserWarning
c081199 [Francesc Alted] Relax the exact message for the ImportError
73f0319 [Francesc Alted] numexpr requisite raised to 2.4.6
0d4ab9a [Francesc Alted] Restored the old numexpr version dependencies to adjust for old requirements
c1aae19 [Francesc Alted] Fixed a lint error
7575ba2 [Francesc Alted] Using constants instead of literals for numexpr version
7a275ce [Francesc Alted] Fixed a typo
93f54aa [Francesc Alted] numexpr section moved to Other API changes section
3b6e58b [Francesc Alted] Removed recomendation for numexpr 2.6.2
f225598 [Francesc Alted] Updated test_compat for numexpr 2.4.6
8bd4ed1 [Francesc Alted] numexpr 2.4.6 requirement moved to other enhancements section
e45b742 [Francesc Alted] Moved pinned versions in CI folder to 2.4.6
6e12e29 [Francesc Alted] Added a notice on the recommended numexpr version
ac62653 [Francesc Alted] Require numexpr 2.4.6
ab79c54 [Francesc Alted] Require numexpr 2.6.2

jreback committed:
TST: print skipped tests files

xref #15341

Author: Jeff Reback <jeff@reback.net>

Closes #15408 from jreback/skip and squashes the following commits:

547bee6 [Jeff Reback] TST: print skipped tests files

jreback committed:
PERF: high memory in MI

closes #13904

Creates an efficient MultiIndexHashTable in cython.
This allows us to efficiently store a multi-index for fast indexing
(.get_loc() and .get_indexer()), with the current tuple-based
(and gil holding) use of the PyObject Hash Table.

This uses the pandas.tools.hashing routines to hash each of the 'values' of a MI
to a single uint64. So this makes MI more memory friendly and much
more efficient. You get these speedups, because the creation of the
hashtable is now much more efficient.

Author: Jeff Reback <jeff@reback.net>

Closes #15245 from jreback/mi and squashes the following commits:

7df6c34 [Jeff Reback] PERF: high memory in MI

jreback committed:
STYLE: flake8 upgraded to 3.3 on conda (#15412)

fixes for E305, 2 blank lines after a class definition

jreback committed:
DOC: use shared_docs for Index.get_indexer, get_indexer_non_unique (#15411)

* STYLE: flake8 upgraded to 3.3 on conda

fixes for E305, 2 blank lines after a class definition

* DOC: use shared_docs for Index.get_indexer, get_indexer_non_unique

fix non-populated doc-strings for some methods in Index (take)

jreback committed:
BLD: use latest conda version with latest miniconda installer on appveyor

change 3.6 build to use numpy=1.12 & add back xlwt (was not on
defaults for a while)

Author: Jeff Reback <jeff@reback.net>

Closes #15415 from jreback/appveyor and squashes the following commits:

2019f37 [Jeff Reback] force numpy version
f82877b [Jeff Reback] remove extra conda list
3ace9f2 [Jeff Reback] CI: use numpy=1.12 on appveyor
6855a7b [Jeff Reback] BLD: use latest conda version with latest miniconda installer on appveyor

jreback committed:
TST: convert yield based test_pickle.py to parametrized to remove warnings

xref #15341

Author: Jeff Reback <jeff@reback.net>

Closes #15416 from jreback/warn and squashes the following commits:

a6af576 [Jeff Reback] TST: convert yield based test_pickle.py to parametrized to remove warnings xref #15341

QuLogic committed:
TST: Parametrize simple yield tests

xref #15341

Author: Elliott Sales de Andrade <quantum.analyst@gmail.com>

Closes #15406 from QuLogic/pytest-simple-yield and squashes the following commits:

b002752 [Elliott Sales de Andrade] TST: Set PYTHONHASHSEED so xdist doesn't break.
8368772 [Elliott Sales de Andrade] TST: Use fixtures for engine/parser where possible.
c6cd346 [Elliott Sales de Andrade] TST: Parametrize remaining simple yield tests.
47bf1a1 [Elliott Sales de Andrade] TST: Replace ENGINES_PARSERS by parametrize.

Commit:
BUG: Ensure the right values are set in SeriesGroupBy.nunique

closes #13453

Author: Diego Fernandez <difernan@redhat.com>

Closes #15418 from aiguofer/gh_13453 and squashes the following commits:

c53bd70 [Diego Fernandez] Add test for #13453 in test_resample and add note to whatsnew
0daab80 [Diego Fernandez] Ensure the right values are set in SeriesGroupBy.nunique

abaldenko committed:
BUG: Concat with inner join and empty DataFrame

closes #15328

Author: abaldenko <abaldenko@massmutual.com>

Closes #15397 from abaldenko/concat_empty_dataframe and squashes the following commits:

47c8735 [abaldenko] BUG: Concat with inner join and empty DataFrame
fc473b7 [abaldenko] BUG: Concat with inner join and empty DataFrame
b86dcb6 [abaldenko] BUG: Concat with inner join and empty DataFrame

jeffcarey committed:
ENH: Added ability to freeze panes from DataFrame.to_excel() (#15160)

closes #15160

Author: Jeff Carey <jeff.carey@gmail.com>

Closes #15291 from jeffcarey/enh-15160 and squashes the following commits:

cef8fce [Jeff Carey] ENH: Added ability to freeze panes from DataFrame.to_excel()

jreback commented:
pls update when you can

jeffcarey committed:
Documents touch-up for DataFrame.to_excel() freeze_panes option (#15436)

redbullpeter committed:
BUG: to_sql convert index name to string (#15404) (#15423)

* Converted index name to string to fix issue #15404 - BUG: to_sql errors with numeric index name - needs conversion to string

* Additional int to string conversion added. Associated test cases added.

* PEP 8 compliance edits

* Removed extraneous brackets

jorisvandenbossche committed:
DOC: add whatsnew for #15423

jreback committed:
TST: remove yielding tests from test_msgpacks.py (#15427)

QuLogic committed:
ENH: Don't add rowspan/colspan if it's 1.

Just a small thing I noticed in a [footnote
here](https://danluu.com/web-bloat/#appendix-irony). Probably can't do
much about the extra classes, but rowspan/colspan seem like easy fixes
to save a few bytes per row/col and it's already done in the other
code path.

Author: Elliott Sales de Andrade <quantum.analyst@gmail.com>

Closes #15403 from QuLogic/no-extra-span and squashes the following commits:

9a8fcee [Elliott Sales de Andrade] Don't add rowspan/colspan if it's 1.

jorisvandenbossche committed:
DOC: correct rpy2 examples (GH15142) (#15450)

mroeschke committed:
BUG: rolling not accepting Timedelta-like window args (#15443)

Remove unnecessary pd.Timedelta

jreback committed:
BUG: testing on windows

- we are passing builds which actually have an error
- fix the small dtype issues

Author: Jeff Reback <jeff@reback.net>

Closes #15445 from jreback/windows and squashes the following commits:

a5b7fb3 [Jeff Reback] change integer to power comparisions
eab15c4 [Jeff Reback] don't force remove pandas
cf3b9bd [Jeff Reback] more windows fixing
efe6a76 [Jeff Reback] add cython to build
8194e63 [Jeff Reback] don't use appveyor recipe, just build inplace
e064825 [Jeff Reback] TST: resample dtype issue xref #15418
10d9b26 [Jeff Reback] TST: run windows tests so failures show up in appeveyor

horta committed:
Merge branch 'master' of https://github.com/Horta/pandas

horta committed:
Merge branch 'master' of https://github.com/Horta/pandas

horta committed:
fixed-size arrays for get_index mapping

Horta commented:
`import numpy as np
import pandas as pd
from pandas import Index

from IPython import get_ipython
ipython = get_ipython()

i = Index(Index(np.arange(1000000)).tolist() + Index(np.arange(1000000)).tolist())

ipython.magic("timeit i.get_indexer_non_unique(i[0:1000000])")`

The above test runs in '714 ms'. The implementation regarding the last commit https://github.com/pandas-dev/pandas/pull/15372/commits/bf4b3f585a8d65827bd984c76797c689e72c35fc makes use of fixed-size arrays only, no resize needed.

I think that is the best we can have in terms of data structure. But it can be further improved by, for example, getting rid of PyObject_RichCompare calls due to Python object comparisons. Maybe those Cython fuse types?

horta committed:
dtype=np.int64

horta committed:
empty and zeros with np.int64

horta committed:
as array

jreback commented:
@Horta please rebase on origin/master.

``git rebase -i origin/master`` and ``git push yourremote thisbranch -f``

Horta commented:
@jreback after doing `git rebase -i origin/master` now I got

```
git status
interactive rebase in progress; onto be4a63fe7
Last commands done (6 commands done):
   pick 6afb8c910 wrong exception handling
   pick c7a1e0090 get_indexer_non_unique for orderable indexes
  (see more in file .git/rebase-merge/done)
Next commands to do (9 remaining commands):
   pick c1b657e19 get_indexer_non_unique for orderable indexes
   pick 47f7ce325 C level list
  (use "git rebase --edit-todo" to view and edit)
You are currently editing a commit while rebasing branch 'master' on 'be4a63fe7'.
  (use "git commit --amend" to amend the current commit)
  (use "git rebase --continue" once you are satisfied with your changes)

nothing to commit, working tree clean
```

Not sure what to do now...

jreback commented:
so the issue is you are using your ``master`` branch like a branch which is generally not recommended, so to fix you can do:

```
git checkout yourbranch
git fetch origin/master
git rebase -ii origin/master
```

Horta commented:
I think I need to start a new pull request because I don't know what is going on anymore... (I've tried the above command)

jreback commented:
superseded by #15468
